File: typescript/src/application/usecase/asset/getAssetUsecase.ts
import Decimal from "decimal.js";
import { AccountRepository } from "../../repository/accountRepository";
import { StockSymbol } from "../../../domain/stockSymbol";
import { UserId } from "../../../domain/userId";

export interface GetAssetUsecaseInput {
  userId: UserId;
}

export interface GetAssetStockOutput {
  symbol: StockSymbol;
  amountJpy: Decimal;
}

export interface GetAssetUsecaseOutput {
  cashAmount: Decimal;
  stocks: GetAssetStockOutput[];
}

export class GetAssetUsecaseException extends Error {}
export class UserNotFoundException extends GetAssetUsecaseException {
  constructor() {
    super("user not found");
  }
}

export class GetAssetUsecase {
  constructor(private readonly accountRepository: AccountRepository) {}

  async run(input: GetAssetUsecaseInput): Promise<GetAssetUsecaseOutput> {
    const account = await this.accountRepository.find(input.userId);
    if (account === undefined) {
      throw new UserNotFoundException();
    }
    const stocks = account.stocks.map((e) => ({
      symbol: e.symbol,
      amountJpy: e.amountJpy,
    }));
    return { cashAmount: account.cash, stocks };
  }
}

File: golang/internal/application/usecase/asset/get_asset_usecase.go
package asset

import (
	"errors"

	"folio/codinginterview/internal/application/repository"
	"folio/codinginterview/internal/domain"

	"github.com/shopspring/decimal"
)

var ErrUserNotFound = errors.New("user not found")

type GetAssetStockOutput struct {
	Symbol    domain.StockSymbol
	AmountJpy decimal.Decimal
}

type GetAssetUsecaseInput struct {
	UserId domain.UserId
}

type GetAssetUsecaseOutput struct {
	CashAmount decimal.Decimal
	Stocks     []GetAssetStockOutput
}

type GetAssetUsecase struct {
	accountRepo repository.AccountRepository
}

func NewGetAssetUsecase(accountRepo repository.AccountRepository) *GetAssetUsecase {
	return &GetAssetUsecase{accountRepo: accountRepo}
}

func (u *GetAssetUsecase) Run(input GetAssetUsecaseInput) (GetAssetUsecaseOutput, error) {
	account, err := u.accountRepo.Find(input.UserId)
	if err != nil {
		return GetAssetUsecaseOutput{}, err
	}
	if account == nil {
		return GetAssetUsecaseOutput{}, ErrUserNotFound
	}

	stocks := make([]GetAssetStockOutput, 0, len(account.Stocks))
	for _, s := range account.Stocks {
		stocks = append(stocks, GetAssetStockOutput{Symbol: s.Symbol, AmountJpy: s.AmountJpy})
	}

	return GetAssetUsecaseOutput{CashAmount: account.Cash, Stocks: stocks}, nil
}
